server/src/main/java/org/elasticsearch/cluster/routing/RoutingNode.java

```diff
@@ -68,6 +68,23 @@ public RoutingNode(String nodeId, DiscoveryNode node, ShardRouting... shards) {
         assert invariant();
     }
 
+    private RoutingNode(RoutingNode original) {
+        this.nodeId = original.nodeId;
+        this.node = original.node;
+        this.shards = new LinkedHashMap<>(original.shards);
+        this.relocatingShards = new LinkedHashSet<>(original.relocatingShards);
+        this.initializingShards = new LinkedHashSet<>(original.initializingShards);
+        this.shardsByIndex = new LinkedHashMap<>(original.shardsByIndex.size());
+        for (Map.Entry<Index, LinkedHashSet<ShardRouting>> entry : original.shardsByIndex.entrySet()) {
+            shardsByIndex.put(entry.getKey(), new LinkedHashSet<>(entry.getValue()));
+        }
+        assert invariant();
+    }
+
+    public RoutingNode copy() {
```

Comment:
Can this be package private instead?

Also, perhaps rename to `mutableCopy`?

Reply:
Sure, though these instances are all mutable to begin with, so maybe leave the name? :)